### PR TITLE
Fix Device-Flow client validation

### PR DESCRIPTION
**Pull request by ChamathNS**
### Proposed changes in this pull request

Fix https://github.com/wso2/product-is/issues/7275

### CODE REVIEW

**Comment by madurangasiriwardena** [comment]
Hi @ChamathNS, travis build is failing with below error message.
`[ERROR] Failed to execute goal org.jacoco:jacoco-maven-plugin:0.7.9:instrument (default-instrument) on project org.wso2.carbon.identity.oauth.endpoint: Unable to instrument file. Error while instrumenting class /home/travis/build/wso2-extensions/identity-inbound-auth-oauth/components/org.wso2.carbon.identity.oauth.endpoint/target/classes/org/wso2/carbon/identity/oauth/endpoint/ciba/CibaAuthResponseHandler.class. 6 -> [Help 1]`

can you check the reason.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/device/DeviceEndpoint.java [hunk 1]
```diff
@@ -57,44 +66,81 @@ public void setDeviceAuthService(DeviceAuthService deviceAuthService) {
...
+
+        if (OAuth2ErrorCodes.INVALID_CLIENT.equals(oAuthClientAuthnContext.getErrorMessage())) {
+            return handleBasicAuthFailure(oAuthClientAuthnContext.getErrorCode(),
+                    oAuthClientAuthnContext.getErrorMessage());
+        } else if (OAuth2ErrorCodes.SERVER_ERROR.equals(oAuthClientAuthnContext.getErrorMessage())) {
 
-        return deviceAuthService.validateClientInfo(clientId);
+            return handleServerError();
+        } else {
+            // Otherwise send back HTTP 400 Status Code
```
Comment (pamodaaw) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/device/DeviceEndpoint.java]: Missing fullstop

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/device/DeviceEndpoint.java [hunk 2]
```diff
@@ -57,44 +66,81 @@ public void setDeviceAuthService(DeviceAuthService deviceAuthService) {
...
+
+    private Response handleBasicAuthFailure(String errorCode, String errorMessage) throws OAuthSystemException {
+
+        OAuthResponse response = OAuthASResponse.errorResponse(HttpServletResponse.SC_UNAUTHORIZED)
+                .setError(OAuth2ErrorCodes.INVALID_CLIENT)
+                .setErrorDescription("Client Authentication failed.").buildJSONMessage();
+        return Response.status(response.getResponseStatus())
+                .header(OAuthConstants.HTTP_RESP_HEADER_AUTHENTICATE, EndpointUtil.getRealmInfo())
+                .entity(response.getBody()).build();
+    }
```
Comment (pamodaaw) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/device/DeviceEndpoint.java]: keep a new line

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/device/DeviceEndpoint.java [hunk 3]
```diff
@@ -57,44 +66,81 @@ public void setDeviceAuthService(DeviceAuthService deviceAuthService) {
...
+
+    private Response handleServerError() throws OAuthSystemException {
+
+        OAuthResponse response = OAuthASResponse.errorResponse(HttpServletResponse.SC_INTERNAL_SERVER_ERROR).
+                setError(OAuth2ErrorCodes.SERVER_ERROR).setErrorDescription("Internal Server Error.")
+                .buildJSONMessage();
+
+        return Response.status(response.getResponseStatus()).header(OAuthConstants.HTTP_RESP_HEADER_AUTHENTICATE,
+                EndpointUtil.getRealmInfo()).entity(response.getBody()).build();
+
```
Comment (pamodaaw) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/device/DeviceEndpoint.java]: Remove unnecessary new line

File: components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/device/DeviceEndpointTest.java [hunk 4]
```diff
@@ -113,11 +124,16 @@ public void testStringValueInSeconds(long value) throws Exception {
 
     @DataProvider(name = "dataValues")
     public Object[][] dataValues() {
+        MultivaluedMap<String, String> mapWithClientId = new MultivaluedHashMap<>();
```
Comment (pamodaaw) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/device/DeviceEndpointTest.java]: Keep a new line after the method declaration

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/device/DeviceEndpoint.java [hunk 5]
```diff
@@ -57,42 +66,79 @@ public void setDeviceAuthService(DeviceAuthService deviceAuthService) {
...
+        OAuthClientAuthnContext oAuthClientAuthnContext;
+        Object oauthClientAuthnContextObj = request.getAttribute(OAuthConstants.CLIENT_AUTHN_CONTEXT);
+        if (oauthClientAuthnContextObj instanceof OAuthClientAuthnContext) {
+            oAuthClientAuthnContext = (OAuthClientAuthnContext) oauthClientAuthnContextObj;
+        } else {
+            oAuthClientAuthnContext = new OAuthClientAuthnContext();
+            oAuthClientAuthnContext.setAuthenticated(false);
+            oAuthClientAuthnContext.setErrorMessage("Client Authentication Failed");
+            oAuthClientAuthnContext.setErrorCode(OAuthError.TokenResponse.INVALID_REQUEST);
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/device/DeviceEndpoint.java:119]: Is invalid request the appropriate error code here?

**Pull request merged**